fns-hl7-pipeline/fn-hl7-json-lake/src/main/kotlin/gov/cdc/dex/hl7/Function.kt

```diff
@@ -68,66 +68,67 @@ class Function {
-                        with(outData) {
-                            jsonlakeOkOutput.value = ok
-                            jsonlakeErrOutput.value = err
-                            cosmosOutput.value = cosmo
+                            outData
+                        )
+                        context.logger.info("Processed OK for HL7 JSON Lake messageUUID: $messageUUID, filePath: $filePath")
+                        if (messageIndex == messages.lastIndex) {
+                            return inputEvent
                         }
-                        return inputEvent
-                    }
-                } catch (e: Exception) {
-                    context.logger.severe("Exception: Unable to process Message messageUUID: $messageUUID, filePath: $filePath, due to exception: ${e.message}")
-                    //publishing the message  to the eventhubSendErrsName topic using EventHub
-                    updateMetadataAndDeliver(startTime, metadata, PROCESS_STATUS_EXCEPTION,
+                    } catch (e: Exception) {
+                        context.logger.severe("Exception: Unable to process Message messageUUID: $messageUUID, filePath: $filePath, due to exception: ${e.message}")
+                        //publishing the message  to the eventhubSendErrsName topic using EventHub
+                        updateMetadataAndDeliver(
+                            startTime, metadata, PROCESS_STATUS_EXCEPTION,
                             null, eventHubMD[messageIndex], gsonWithNullsOn,
                             inputEvent, e,
                             listOf(FunctionConfig.PROFILE_FILE_PATH),
-                            outData)
-                    context.logger.info("Processed ERROR for HL7 JSON Lake messageUUID: $messageUUID, filePath: $filePath")
-                    with(outData) {
-                        jsonlakeOkOutput.value = ok
-                        jsonlakeErrOutput.value = err
-                        cosmosOutput.value = cosmo
-                    }
-                    return inputEvent
-                } // .catch
-
-            } catch (e: Exception) {
-                // message is bad, can't extract fields based on schema expected
-                context.logger.severe("Unable to process Message due to exception: ${e.message}")
-                e.printStackTrace()
-                return JsonObject()
+                            outData
+                        )
+                        context.logger.info("Processed ERROR for HL7 JSON Lake messageUUID: $messageUUID, filePath: $filePath")
+                        return inputEvent
+                    } // .catch
 
-            } // .catch
+                } catch (e: Exception) {
+                    // message is bad, can't extract fields based on schema expected
+                    context.logger.severe("Unable to process Message due to exception: ${e.message}")
+                    e.printStackTrace()
+                    return JsonObject()
```

Comment:
Marcello, I did update to comply getting the exception in the error queue and it will go in the cosmos db as well.
I was not sure whether to do only the error queue. If we leave it out of the cosmos db, we create a precedent where some messages are in error and not in cosmos. Maybe later this will change.
Another thing is that this message is basically unidentifiable, since the catch is the outside one,  and as you said
the message will not make it that far,
Let me know.

Reply:
that's is perfect emil - saving to cosmos Db is the right thing! 
again, we don't know what kind of situations will trigger this (prob, never :)) but it's better to have some visibility than none.